feat(file-exclusion): Add option to exclude files from mutation run

michielpeeters opened the pull request:
1. Added option in CLI -fte or --files-to-exclude for capabilities to exclude files
2. Added option for JSON file that an array with files to exclude can be used
3. Updated the tests that correspond with this change
4. Updated readme.md to reflect the new changes of the stryker json file

Sorry for the late response of this issue. Aim to resolve #34.

Mobrockers commented on src/Stryker.CLI/Stryker.CLI/OptionsBuilder.cs:
```diff
@@ -59,11 +65,17 @@ public class OptionsBuilder
                 {   
                     return config.GetSection(thresholdOptionsSectionKey).GetValue<T>(defaultValue.JsonKey);
                 }
+                else if (config.GetSection("files-to-exclude").Exists() &&
```
The configuration section should not have been loaded in this method but since it has already slipped through with a previous pr i'll allow it here and create a new issue to find a better place for both files-to-include and threshold-options

michielpeeters replied:
Yeah, I thought already that shouldn't be there. But due to the fact that the CLI and the stryker options file are merged it is rather difficult to split them in a nicer way. When accepted I'll search for a solution.

Mobrockers replied:
I created #203 to solve this issue after your PR has been merged.

michielpeeters replied:
Ok cool, I'll resolve it for now.

Mobrockers commented on src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs:
```diff
@@ -12,6 +12,8 @@
 
 namespace Stryker.CLI.UnitTest
 {
+    using System.Linq;
```
We place usings outside of the namespace

michielpeeters replied:
Ahhhh crap, sorry it's my resharper file that is bugging! Let me change it quickly ;-)

Mobrockers replied:
No worries :)

michielpeeters replied:
Fixed!

michielpeeters commented on src/Stryker.CLI/Stryker.CLI/OptionsBuilder.cs:
```diff
@@ -6,6 +6,9 @@
 using System.Linq;
 namespace Stryker.CLI
 {
+    using System.Collections.Generic;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InitialisationProcessTests.cs:
```diff
@@ -12,6 +12,8 @@
 
 namespace Stryker.Core.UnitTest.Initialisation
 {
+    using System.Collections.Generic;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InputFileResolverTests.cs:
```diff
@@ -10,6 +10,8 @@
 
 namespace Stryker.Core.UnitTest.Initialisation
 {
+    using Options;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:
```diff
@@ -9,9 +9,11 @@
 
 namespace Stryker.Core.Initialisation
 {
+    using Options;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs:
```diff
@@ -18,6 +18,8 @@
 
 namespace Stryker.Core.MutationTest
 {
+    using Options;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -6,6 +6,8 @@
 
 namespace Stryker.Core.Options
 {
+    using Newtonsoft.Json;
```
Fixed!

michielpeeters commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InputFileResolverTests.cs:
```diff
@@ -141,12 +143,56 @@ public void InputFileResolver_InitializeShouldIgnoreBinFolder(string folderName)
             result.ProjectContents.Children.Count.ShouldBe(1);
         }
 
+        [SkippableFact]
```
Yes, due to the fact that windows statements are used and all the other tests are also skippablefact :-). I think that they are there for a reason isn't it?

Mobrockers replied:
We currently have a PR open to resolve the skippable tests, making them work on platforms other than Windows, so it would be nice to not introduce a new windows-only test. What windows only features are used?

michielpeeters replied:
Ah! #197 you mean? Good point, to be honest I saw these tests and I extended it for my specific testing purposes. I can't actually tell you what kind of specific windows feature is being used. Do you want that I use already the basic setup of the PR 197 or will the PR suppliers change this accordingly?

Mobrockers replied:
#197 only needs a small fix before it can be merged so I don't expect that to take long. If this PR is merged sooner though then I suppose it would be on them to fix your test. I would prefer if it was fixed in this PR ;) I suggest you work on the other comments first and see how it goes.

michielpeeters commented on src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs:
```diff
@@ -215,5 +218,35 @@ public void StrykerCLI_OnMutationScoreAboveThresholdBreak_ShouldReturnExitCode0(
+
+            var target = new StrykerCLI(mock.Object);
+
+            target.Run(new[] { argName, "['/StartUp.cs']" });
```
Agreed, Only I was thinking, now I have a JSON format that you can apply e.g. '['C:\\ExampleProject\\StartUp.cs','C:\\ExampleProject\\Recursive.cs']' but that looks ugly. Maybe it is a better idea to do something like this 'C:\\ExampleProject\\StartUp.cs,C:\\ExampleProject\\Recursive.cs'? It is still something what is bugging me and I can't think about a better idea how to provide it.

Mobrockers replied:
I think `['C:\ExampleProject\StartUp.cs','C:\ExampleProject\Recursive.cs']` is acceptable so long as it can be one line. For bigger projects with more excludes the file config should be used anyway.

michielpeeters replied:
True, I'll update everything to JSON input. In the new issue I'll try to think about of how to resolve it in a more easy and fancy way :-).

michielpeeters replied:
I have to admit that providing a JSON string through a unit test is quite a challenge. Somehow the Newtonsoft Serializer has really big trouble translating the input. I have it solved but it needs some rework with fetching data which you have already stated in the issue #203 and it looks really yuck ... I've committed the changes so you can review them.

Mobrockers commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -88,5 +94,20 @@ private ThresholdOptions ValidateThresholds(int thresholdHigh, int thresholdLow,
+            var excludedFiles = new List<string>();
+            try
+            {
+                excludedFiles = JsonConvert.DeserializeObject<List<string>>(filesToExclude);
```
It would be nice if we could warn users on non-existant files but I don't think we should mark it as an invalid configuration. Logging as warning should be fine.

Mobrockers commented:
#197 has been merged so you will need to resolve some conflicts in the files they touched (namely the InputFileResolver where they fixed the cross-platform issue we had with our tests)

richardwerkman commented:
@michielpeeters Do you have time to resolve the conflicts and look at our feedback?

michielpeeters commented:
@richardwerkman  Certainly! I only needed to find some time to process the feedback sorry for the delay!

michielpeeters commented:
@richardwerkman @Mobrockers got everything resolved, fixed code conflicts and added greater support for relative paths :-). You will find a todo in the code because I haven't anything available to report back to the console or some sort of logging mechanism. I have some ideas regarding issue #203 where I can resolve this as well, but I need to rebuild some stuff then.

michielpeeters commented:
Ok, this is weird. Locally my tests run successful, but when using the build server it brakes. I know it has to do something with the FilesToExclude checks but it's now difficult to pinpoint where it goes wrong.

michielpeeters commented:
@richardwerkman @Mobrockers You guys have any chance by reviewing the latest code? And do we, as contributes, have the ability to trigger build on the Azure Build Pipeline? Thx guys I really want to start with a new issue ;-) ...

michielpeeters commented on src/Stryker.CLI/README.md:
```diff
@@ -68,9 +68,16 @@ If you want to decide on your own mutation score thresholds, you can configure t
+1. When using the relative path option ./ (for Windows .\) it uses the current directory where it is scanning for files (The final directory is: C:\ExampleDirectory\OtherDirectory\ExampleClass.cs).
+2. When using the relative path option ../ (for Windows ..\) it moves up one directory that is being scanned (The final directory is: C:\ExampleDirectory\ExampleClass.cs).
 
 The only format which is currently accepted is a JSON formatted string array. If an invalid JSON string is provided a ValidationException is thrown directing to you to provide a correct JSON formatted string. 
```
Good point, I was still thinking as a framework guru lol ... will change it.

michielpeeters commented on src/Stryker.CLI/README.md:
```diff
@@ -68,9 +68,16 @@ If you want to decide on your own mutation score thresholds, you can configure t
+The following rules are used when using relative paths. For these rules the following example directory is used: C:\ExampleDirectory\OtherDirectory\
+
+1. When using the relative path option ./ (for Windows .\) it uses the current directory where it is scanning for files (The final directory is: C:\ExampleDirectory\OtherDirectory\ExampleClass.cs).
+2. When using the relative path option ../ (for Windows ..\) it moves up one directory that is being scanned (The final directory is: C:\ExampleDirectory\ExampleClass.cs).
```
Well, I have added this due to the fact that most of the times (at least I have that problem :P) that when using relative paths that people don't know the difference. By this it gives more details of how to use the exclude files option :-). I can remove it though.

richardwerkman replied:
Well, I'm now thinking that maybe we should handle paths platform unspecific. For example Stryker is configured for a team of 10 developers. Some might work on windows and others on mac. We would want Stryker to exclude the files on all platforms with the same config file. Could we achieve this? Maybe by replacing / to `Environment.PathSeperator` ?

This would also allow us to document this feature platform unspecific. That would be much nicer in my opinion!

michielpeeters replied:
Finally I got a local windows ubuntu system running :-) so I can finally try to resolve this. I have discovered when using absolute paths C:\ExampleDirectory\Test.cs and when parsing it on ubuntu it gives a full path /someawesomedirectory/C:\ExampleDirectory\Test.cs which is ofcourse correct due to the fact that it tries to make use of the directory seperator where the tests are currently executed. With that said I think that absolute paths shouldn't be used and only accept relative paths. Otherwise it is not possible (yes everything is possible, but it needs some code which removes the filesystemroot) to create a config file that can be used by multiple environments/members. What do you think @richardwerkman?

michielpeeters replied:
@richardwerkman I've updated the code. It now determines which kind of separator is being used and replaces it accordingly with the platform separator.

michielpeeters commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -95,12 +95,24 @@ private ThresholdOptions ValidateThresholds(int thresholdHigh, int thresholdLow,
+
+                    if (!File.Exists(fullPath))
+                    {
+                        // TODO: Write something to the console that the provided excluded file does not exist.
```
Ah, nice! Will change it!

richardwerkman commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -109,20 +109,26 @@ private List<string> ValidateFilesToExclude(string basePath, string filesToExclu
             return excludedFiles;
         }
+
+        private static string GetPlatformSupportedFilePath(string excludedFile)
```
This could be usefull elsewhere too. Maybe the [FilePathUtils](https://github.com/stryker-mutator/stryker-net/blob/master/src/Stryker.Core/Stryker.Core/FilePathUtils.cs) is a good place for this?

richardwerkman commented on src/Stryker.CLI/README.md:
```diff
@@ -67,6 +67,15 @@ If you want to decide on your own mutation score thresholds, you can configure t
+It is advisable to use relative paths. When someone else is running Stryker for you unit test project, paths are automatically resolved. It is still possible to use absolute paths to files.
+
+The only format which is accepted is a JSON formatted string array.
+When you want to exclude multiple files (for example more then 5) it is better to use the stryker configuration file because it is more easier to handle multiple files.
```
```suggestion
When you want to exclude multiple files (for example more then 5) it is advised to use the stryker configuration file because it is easier to handle multiple files.
```

richardwerkman commented on src/Stryker.CLI/README.md:
```diff
@@ -67,6 +67,15 @@ If you want to decide on your own mutation score thresholds, you can configure t
+
+`dotnet stryker --files-to-exclude ['./ExampleClass.cs','./ExampleDirectory/ExampleClass2.cs']`
+
+It is advisable to use relative paths. When someone else is running Stryker for you unit test project, paths are automatically resolved. It is still possible to use absolute paths to files.
```
```suggestion
We recommend to use relative paths. When someone else is running Stryker for you unit test project, paths are automatically resolved. It is also possible to use absolute paths to files.
```

richardwerkman commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InputFileResolverTests.cs:
```diff
@@ -136,17 +134,60 @@ public void InputFileResolver_InitializeShouldIgnoreBinFolder(string folderName)
+            // the bin, obj and node_modules folders should be skipped
+            string sourceFile = File.ReadAllText(_currentDirectory + "/Initialisation/TestResources/ExampleSourceFile.cs");
+            string projectFile = @"
+<Project Sdk=""Microsoft.NET.Sdk"">
```
I've seen this exact string more than once in this test class. It would improve much if this default string could be shared across the class.

richardwerkman commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -88,5 +99,38 @@ private ThresholdOptions ValidateThresholds(int thresholdHigh, int thresholdLow,
+                    if (!File.Exists(fullPath))
+                        _logger.LogWarning($"The specified file to exclude {fullPath} could not be found. Did you mean to exclude another file?");
+
+                    excludedFiles.Add(fullPath);
```
```suggestion
                    excludedFiles.Add(fullPath);
                    _logger.LogInformation($"File {0} will be excluded from mutation", fullPath);
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -88,5 +99,38 @@ private ThresholdOptions ValidateThresholds(int thresholdHigh, int thresholdLow,
+                    var fullPath = Path.GetFullPath(Path.Combine(basePath, platformFilePath));
+
+                    if (!File.Exists(fullPath))
+                        _logger.LogWarning($"The specified file to exclude {fullPath} could not be found. Did you mean to exclude another file?");
```
```suggestion
                        _logger.LogWarning($"The specified file to exclude {0} could not be found. Did you mean to exclude another file?", fullPath);
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/Options/StrykerOptions.cs:
```diff
@@ -88,5 +99,38 @@ private ThresholdOptions ValidateThresholds(int thresholdHigh, int thresholdLow,
+        {
+            return excludedFile.Replace("\\", Path.DirectorySeparatorChar.ToString()).Replace("/", Path.DirectorySeparatorChar.ToString());
+
+            //return excludedFile.Replace(excludedFile.Contains("\\") ? "\\" : "/", Path.DirectorySeparatorChar.ToString());
```
Remove commented line

richardwerkman commented on src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:
```diff
@@ -62,25 +62,28 @@ public ProjectInfo ResolveInput(string currentDirectory, string projectName)
             }
-            foreach (var file in _fileSystem.Directory.GetFiles(path, "*.cs", SearchOption.TopDirectoryOnly))
+
+            foreach (var file in _fileSystem.Directory.GetFiles(path, "*.cs", SearchOption.TopDirectoryOnly).Except(filesToExclude))
```
I have not tested this yet but I suddenly doubt this will work. Later in the process we feed the Roslyn compiler the abstract syntax trees of all files. Roslyn needs this to compile correctly. I guess that if an important file is excluded Stryker cannot compile the program because of missing types.

I think it's better to read the file and add it to the composite. But we mark it as excluded, to skip it later from exclusion. This also has an advantage. We could display in the final console report that this file was excluded. Using this current method we can't do that.

michielpeeters replied:
Hmmm, it's still getting heavier this pull request :-D ... haha but I understand what you mean. Quite funny though, due to the fact that you have in an earlier comment suggested to do it like this :-). Anyway I'll be going back to the drawing board!

richardwerkman replied:
Yeah I know I'm sorry about that. I was just looking at your code style the first time. Now I took some time to look at the functionallities. I like how you are so motivated to get this right! The feature would be more awesome this way too if you ask me :D

michielpeeters replied:
No sorry man! Just getting the job done :-). I've convinced many people to use Stryker.net so yea ... it should be good :-)

michielpeeters replied:
@richardwerkman Aight, got it :-), you can check it out!

richardwerkman commented on src/Stryker.CLI/Stryker.CLI.UnitTest/StrykerCLITests.cs:
```diff
@@ -215,5 +229,41 @@ public void StrykerCLI_OnMutationScoreAboveThresholdBreak_ShouldReturnExitCode0(
+            var secondFileToExclude = Path.Combine(_currentDirectory, "ExampleDirectory", "Recursive.cs");
+            var thirdFileToExclude = Path.Combine(_currentDirectory, "ExampleDirectory", "Recursive2.cs");
+
+            mock.Verify(x => x.RunMutationTest(It.Is<StrykerOptions>(o =>
```
You could use shouldly to verify the array nicer.

michielpeeters replied:
Oh, nice! Didn't know of that one :-)!

richardwerkman commented:
While testing I notices a fault in the architecture for this feature. We can't know what files to exclude when creating the `StrykerOptions` because the "projectUnderTest" has not been resolved yet. 

By moving the logic to the "InputFileResolver" I managed to succesfull exclude a file. Since I had already fixed it I pushed it to your branch.

michielpeeters commented on src/Stryker.Core/Stryker.Core/Reporters/ConsoleReportReporter.cs:
```diff
@@ -29,8 +29,12 @@ public void OnMutantsCreated(IReadOnlyInputComponent inputComponent)
             Console.WriteLine("");
 
-            _chalk.Default($"{inputComponent.TotalMutants.Count()} mutants have been created. Each mutant will now be tested, this could take a while. {Environment.NewLine}");
+            _chalk.DarkGray($"{inputComponent.ExcludedFiles.Count} have been excluded. For these files there are no mutants created. {Environment.NewLine}");
```
You mean the OnAllMutantsTested process or?

michielpeeters commented:
Thanks @richardwerkman! I was already working on it :-D ... anyhow I'll fix the tests and still searching of generating a good ending report what you have suggested. Sorry for no/late response.

richardwerkman commented:
@michielpeeters You're welcome! This was one big merge but I think it went okay. We appreciate your effort of course so if you need any help please let us know.

richardwerkman commented:
@michielpeeters It would be nice if we could close this PR anytime soon, since the conflicts just keep coming. If you don't have the time to finnish this please let us know so we can take it off your hands.

michielpeeters commented:
Hi @richardwerkman I really want to finish this PR but struggling to find the time to finish it. A lot is going on at the moment. I'm not throwing the towel into the ring (I guess that's bad english :P). So hopefully I can still contribute in the near future. If you have the chance the finish it please do!

Mobrockers commented on src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:
```diff
@@ -87,9 +88,8 @@ private void RunStryker(StrykerOptions options)
-        private void HandleBreakingThresholdScore(StrykerOptions options, StrykerRunResult results)
-        {
-            _logger.LogError($@"Final mutation score: {results.MutationScore} under breaking threshold value {options.ThresholdOptions.ThresholdBreak}.
+        private void HandleBreakingThresholdScore(StrykerOptions options, StrykerRunResult results) {
```
Brace should be on new line

Mobrockers commented on src/Stryker.CLI/Stryker.CLI/OptionsBuilder.cs:
```diff
@@ -35,59 +37,78 @@ public class OptionsBuilder
+        private T GetOption<V, T>(V cliValue, CLIOption<T> option)
         {
-            if (value != null)
+            if (option.ValueType == CommandOptionType.NoValue && cliValue is string cliNoValueValue && cliNoValueValue == "on")
```
Why can't we simply ask for HasValue() on a NoValue type instead of parsing the string value?

Mobrockers commented on src/Stryker.Core/Stryker.Core/FilePathUtils.cs:
```diff
@@ -14,4 +14,11 @@ public static string ConvertPathSeparators(string filePath)
 		}
 	}
+
+    public static string ConvertToPlatformSupportedFilePath(string filePath)
```
Does this not do the same as ConvertPathSeperators?
